include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl

```diff
@@ -3,428 +3,573 @@
 
 		if (highMsb == -1)
 		{
 			const uint32_t lowBits = uint32_t(val);
-			const int32_t lowMsb = find_msb_helper<uint32_t>::findMSB(lowBits);
+			const int32_t lowMsb = find_msb_helper<uint32_t>::__call(lowBits);
 			if (lowMsb == -1)
 				return -1;
 
 			return lowMsb;
 		}
 
 		return highMsb + 32;
-#else
-		return glm::findMSB(val);
-#endif
 	}
 };
+template<typename UInt64> NBL_PARTIAL_REQ_TOP(is_same_v<UInt64, uint64_t>)
+struct find_lsb_helper<UInt64 NBL_PARTIAL_REQ_BOT(is_same_v<UInt64, uint64_t>) >
+{
+	static int32_t __call(NBL_CONST_REF_ARG(uint64_t) val)
+	{
+		const uint32_t lowBits = uint32_t(val);
+		const int32_t lowLsb = find_lsb_helper<uint32_t>::__call(lowBits);
+
+		if (lowLsb == -1)
+		{
+			const uint32_t highBits = uint32_t(val >> 32);
+			const int32_t highLsb = find_lsb_helper<uint32_t>::__call(highBits);
+			if (highLsb == -1)
+				return -1;
+			else
+				return 32 + highLsb;
+		}
+
+		return lowLsb;
+	}
+};
+
+#define AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER_2_ARG_FUNC(HELPER_NAME, SPIRV_FUNCTION_NAME, RETURN_TYPE)\
+template<typename T> NBL_PARTIAL_REQ_TOP(always_true<decltype(spirv::SPIRV_FUNCTION_NAME<T>(experimental::declval<T>()))>)\
+struct HELPER_NAME<T NBL_PARTIAL_REQ_BOT(always_true<decltype(spirv::SPIRV_FUNCTION_NAME<T>(experimental::declval<T>()))>) >\
+{\
+	using return_t = RETURN_TYPE;\
+	static inline return_t __call(const T a, const T b)\
+	{\
+		return spirv::SPIRV_FUNCTION_NAME<T>(a, b);\
+	}\
+};
 
-template<int N>
-struct find_msb_helper<vector<uint32_t, N> >
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER_2_ARG_FUNC(max_helper, fMax, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER_2_ARG_FUNC(max_helper, uMax, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER_2_ARG_FUNC(max_helper, sMax, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER_2_ARG_FUNC(min_helper, fMax, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER_2_ARG_FUNC(min_helper, uMax, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER_2_ARG_FUNC(min_helper, sMax, T)
+
+#else // C++ only specializations
+
+template<typename T>
+requires concepts::Scalar<T>
+struct clamp_helper<T, T>
 {
-	static vector<int32_t, N> findMSB(NBL_CONST_REF_ARG(vector<uint32_t, N>) val)
+	using return_t = T;
+	static inline return_t __call(const T val, const T min, const T max)
 	{
-#ifdef __HLSL_VERSION
-		return spirv::findUMsb(val);
-#else
-		return glm::findMSB(val);
-#endif
+		return std::clamp(val, min, max);
+	}
+};
+template<typename T>
+requires concepts::IntegralScalar<T>
+struct bitReverse_helper<T>
+{
+	static inline T __call(NBL_CONST_REF_ARG(T) arg)
+	{
+		return glm::bitfieldReverse(arg);
 	}
 };
+template<typename Matrix>
+requires concepts::Matrix<Matrix>
+struct transpose_helper<Matrix>
+{
+	using transposed_t = typename matrix_traits<Matrix>::transposed_type;
 
-template<int N>
-struct find_msb_helper<vector<int32_t, N> >
+	static transposed_t __call(NBL_CONST_REF_ARG(Matrix) m)
+	{
+		return reinterpret_cast<transposed_t&>(glm::transpose(reinterpret_cast<typename Matrix::Base const&>(m)));
```

Comment:
always call GLM with explicit template params

Reply:
whats the problem with explicit template params for `glm::transpose` ?

Reply:
"no matching overloaded function found" so i left it as it is since it works

Reply:
yeah because the signature of `glm::tranpose` is  `template<length_t C, length_t R, typename T, qualifier Q>`
https://github.com/g-truc/glm/blob/33b4a621a697a305bc3a7610d290677b96beb181/glm/matrix.hpp#L131

use the traits to obtain the matrix's C R and T (remember C and R are kinda flipped in our alias)

Q you can't really deduce

Reply:
is it safe then to assume qualifier is always `glm::qualifier::highp`? our matrices always use `glm::qualifier::highp`

Reply:
> is it safe then to assume qualifier is always `glm::qualifier::highp`? our matrices always use `glm::qualifier::highp`

that last template parameter you can let the compiler deduce